Update content area of login screens to use StepContainerV2

tellthemachines opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #102039

## Proposed Changes

* Updates the login screens using `isWhiteLogin` design to make use of the `Step.CenteredColumnLayout` wireframe, including rendering titles with `Step.Heading`.
* Removes Gravatar and GravPoweredClient login screens from `isWhiteLogin` condition (they were already overriding most of the styles anyway)
* Re-adds WP Cloud logo to the WP Cloud login page which _also_ uses `isWhiteLogin`, and which I missed in #102353 😅 

* In order to separate `isWhiteLogin` screens from all the other login screens using the same component, I extracted the header rendering logic into its own file, and further extracted the header _text_ logic into its own function, so that we can use the header text for `isWhiteLogin` screens, and still render all the extra pre- and post-header bits and pieces for all the other screens.


## Testing Instructions

Affects the following screens:

* http://calypso.localhost:3000/log-in/
* Login with existing account: go to  http://calypso.localhost:3000/setup/onboarding, click "Continue with email" and use an email associated with an existing account.
* Automattic partner portal, reached through "https://hosts.automattic.com"
* Wp Cloud: from https://wp.cloud/ click on "Start here"
* Akismet: from https://akismet.com/ click on "Log in"

Content for "log in with existing account" should match the following design:

<img width="900" alt="Screenshot 2025-04-24 at 1 44 17 pm" src="https://github.com/user-attachments/assets/c0cd8de2-b99e-400c-9f6f-d239e4219103" />

We're not quite there yet, mainly due to the notice used. Current state of this PR is:

<img width="1318" alt="Screenshot 2025-05-01 at 2 47 38 pm" src="https://github.com/user-attachments/assets/a2be1be5-913e-4ec9-98f1-0496968b6364" />

There's also an additional notice that sometimes pops up about using different languages:

<img width="1281" alt="Screenshot 2025-05-01 at 2 48 32 pm" src="https://github.com/user-attachments/assets/04482cb6-1524-4514-aa56-35619c389a5b" />

@fcoveram how should we deal with this other notice?

Also, should the existing account notice be using the WP component instead of the Calypso one?


### Other affected screens

<img width="1307" alt="Screenshot 2025-05-01 at 2 47 55 pm" src="https://github.com/user-attachments/assets/362fa369-6817-466e-88ab-deda58cc2b32" />

<img width="1419" alt="Screenshot 2025-05-01 at 2 59 35 pm" src="https://github.com/user-attachments/assets/9926e71d-ede3-4ea7-81f6-2b0e689112a2" />

<img width="1420" alt="Screenshot 2025-05-01 at 2 59 51 pm" src="https://github.com/user-attachments/assets/3f2934fb-0a0a-430b-85f9-e104243600b1" />


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

tellthemachines committed:
Update content area of login screens to use StepContainerV2

tellthemachines committed:
Dead code removal

tellthemachines committed:
Untangle Gravatar and WP Job Manager logins from isWhiteLogin

tellthemachines committed:
Extract header text rendering logic from LoginBlock

tellthemachines committed:
Correct header text logic and wp cloud logo

tellthemachines committed:
Move things around and re-style them

tellthemachines commented on client/blocks/login/login-form.jsx:
```diff
@@ -360,23 +359,6 @@ export class LoginForm extends Component {
 		}
 	}
 
-	renderLoginFromSignupNotice() {
```
Moving this notice outside of the form so we can render it just below the page heading.

tellthemachines commented on client/login/wp-login/index.jsx:
```diff
@@ -575,39 +595,82 @@ export class Login extends Component {
+			>
+				{ this.renderI18nSuggestions() }
+
+				<DocumentHead
```
It feels super weird to have this DocumentHead inside the Main component but I'm guessing removing it would require a much larger refactor 😅

zaguiini replied:
Why exactly? The only dependency of it is the `canonicalUrl`, which is defined right above the component.

tellthemachines replied:
Because JSX is supposed to emulate markup, and in HTML you'd never have the document head inside the main content. The code may work correctly, but it feels horribly unsemantic.

talldan replied:
Not trying to speak on behalf of Luis, but I feel the question may have been "why is it a large refactor?".

tellthemachines replied:
Oh! I noticed we're using `DocumentHead` like this in many places, so if we decided to move it somewhere where it makes more sense semantically we should probably be consistent and do it for all instances.

jsnajdr replied:
For what is worth, Next.js also has a [`Head`](https://nextjs.org/docs/pages/api-reference/components/head) component that does exactly the same thing: set something in the `<head>` element.

Doing it with the proper nesting and "semantically" is probably outright impossible, because React apps never control the entire HTML document, they are always mounted in some sub-element of `<body>`.

tyxla replied:
React 19 does [support all the metadata tags](https://react.dev/blog/2024/12/05/react-19#support-for-metadata-tags) dropped anywhere in your component tree, so hopefully at some point we'll be able to migrate to it 🤞

tellthemachines commented on client/login/wp-login/index.jsx:
```diff
@@ -575,39 +595,82 @@ export class Login extends Component {
+						}
+						heading={ <Step.Heading text={ headerText } /> }
+					>
+						{ mainContent }
```
We could further break down this mainContent to use StepContainer wrappers but that will require taking apart more bits of LoginBlock. It might be better left to a follow-up PR because this diff is already pretty big.

(I'm not 100% sure we _need_ to have a LoginBlock component, because they're both looking at similar conditions to render things. There's no real separation of concerns.)

tellthemachines commented:
Note there are a bunch of other screens using the same Login components that **should not** be affected by these changes. These are the ones I have unearthed so far:

Gravatar

<img width="810" alt="Screenshot 2025-05-01 at 3 30 57 pm" src="https://github.com/user-attachments/assets/257d6a85-6665-47f9-9967-1bbc3f065b7a" />

Gravatar-powered WP Job Manager:

<img width="828" alt="Screenshot 2025-05-01 at 3 31 36 pm" src="https://github.com/user-attachments/assets/4f61b016-9141-4d50-a931-88d9681ae576" />

Jetpack:

<img width="1313" alt="Screenshot 2025-05-01 at 3 32 16 pm" src="https://github.com/user-attachments/assets/04a42d50-d990-42d5-b442-c1fa1038489e" />

Blaze Pro:

<img width="1132" alt="Screenshot 2025-05-01 at 3 32 08 pm" src="https://github.com/user-attachments/assets/97e24f79-9949-4429-99f7-cf07b95a027b" />

Woo:

<img width="1126" alt="Screenshot 2025-05-01 at 3 32 37 pm" src="https://github.com/user-attachments/assets/8946a311-10c4-473c-9c6d-62b8a0a29870" />

(The last two are pretty similar to the `isWhiteLogin` screens but they aren't using that logic)

fcoveram commented:
I don’t know how to run the PR in my local, but when I clicked on the calypso live link, I was able to land on the login page ([`https://container-agitated-matsumoto.calypso.live/log-in`](https://container-agitated-matsumoto.calypso.live/log-in) ). There, I noticed a few things.

The 6-cols container has a side padding that shrinks the area where actions are.

![CleanShot 2025-05-01 at 10 32 39@2x](https://github.com/user-attachments/assets/eafc9613-03a6-4b67-b95e-e9d250707374)

On the smallest breakpoints, the spacing breaks the alignment.

<img src=https://github.com/user-attachments/assets/bcd6490b-bfad-4b77-8462-3de5c3f9f8e4 alt="" width=200>

When changing the viewport from `480px` to `479px` there is a layout change that makes the actions smaller, I wonder if that also happens in other screens that stack buttons.

https://github.com/user-attachments/assets/988aaa7c-0856-439e-b20e-5d8d0294a91d

The whole content area is not vertically aligned to balance the top and bottom spaces according to the browser window, leaving an empty space at the bottom.

![CleanShot 2025-05-01 at 10 43 39@2x](https://github.com/user-attachments/assets/5eaf8545-f302-4b00-9b89-d88e536bd3ee)

I acknowledge the PR scope is not about copy change, yet I want to point out the different in
* ToS and Privacy text. It's different than the mockup.
* Input title. It should follow the sentence case logic.

---

Regarding your questions

> how should we deal with this other (language change) notice?

Pinging @crisbusquets and @nuriapenya in case they know about an strategy for dealing with language change call outs. I remember seeing some tickets mentioning a general revision, but I'm unaware of the outcome.

> …should the existing account notice be using the WP component instead of the Calypso one?

We're working on a new Notice component (dhW7zYBamXJIeJIE3wy5f4-fi-111_840), and on this thread (p1742928013644859/1742810438.892249-slack-CNGQYA3B9), there is an agreement to use an updated version of the Calypso component. But I'm unaware of the code state and the next steps.

zaguiini commented on client/login/wp-login/style.scss:
```diff
@@ -297,6 +297,14 @@ $image-height: 47px;
 		padding: 0;
 	}
 
+	.step-container-v2__content-wrapper {
```
Let's not add StepContainerV2 CSS overrides, please.

Why do we need this extra spacing?

tellthemachines replied:
The design has a bit of space between the top bar and the heading. According to Francisco's comment above, it's meant to be vertically aligned

> The whole content area is not vertically aligned to balance the top and bottom spaces according to the browser window, leaving an empty space at the bottom.

How should I go about doing this using vanilla StepContainerV2?

zaguiini replied:
Can't we rely on the `verticalAlign="center"` prop from the `CenteredColumnLayout` wireframe?

tellthemachines replied:
Ah, that works, thanks!

zaguiini reviewed (dismissed):
I'll return to this later for a deeper review, but I'm asking for changes because introducing StepContainer overrides is the root of all evil, and I don't want it to happen again. 😅

tellthemachines committed:
Use component alignment

tellthemachines committed:
Remove extra padding and update tos

tellthemachines commented:
Thanks for the reviews folks! I think I've addressed all the feedback so far.

> When changing the viewport from `480px` to `479px` there is a layout change that makes the actions smaller, I wonder if that also happens in other screens that stack buttons.

This only happens on `isWhiteLogin` screens (so the ones affected by this PR, not the ones that I listed above as unchanged). The rules that do it are [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/login/wp-login/style.scss#L933). How do you reckon we should improve this?

fcoveram commented:
> This only happens on isWhiteLogin screens (so the ones affected by this PR, not the ones that I listed above as unchanged). The rules that do it are [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/login/wp-login/style.scss#L933). How do you reckon we should improve this?

So if it's a rule for the whole form area, how is that at 482px the input and actions are full width? I don't want us to hack the max-width best practice, but I think that from `599px` to the smallest, the sizes should be the same.

zaguiini reviewed (commented):
I'm noticing some things:

1. I see the login form blink upon refresh (not happening in prod)
2. It's not centered aligned on load. There's a layout shift
3. I don't see the subheader text anymore. Is that expected? We have it in prod.
4. Then I see a massive blank space between my user action and the change account escape hatch
5. Finally, the page scrolls when it shouldn't

https://github.com/user-attachments/assets/a66b2719-3a61-42d1-9954-51586bc909b2

That layout shift happens even when entering the page as a guest:

https://github.com/user-attachments/assets/b024e842-ab51-4d7f-b464-7e8344918e0f

Other than that, it's looking almost perfect. Some style comments:

1. Depending on the screen size, there's some extra margin between the border and the UI elements. Can we make that consistent? By that, I mean always stretch it starting from the mobile breakpoint.
2. The "Lost your password?" link has different styles compared to the "Create an account" one. It would be awesome if we could standardize that too.

https://github.com/user-attachments/assets/c9f64ff3-0410-4512-92d2-b5d51fac739e

## Helpers

### Login form blink

I managed to reproduce it both locally and on Calypso.live. This is caused by SSR, which as you'll read below, doesn't seem to be enabled in production? I don't know.

### Layout shift

This is happening because the `CenteredColumnLayout` wireframes uses the `context.isSmallViewport` variable, which is `false` on SSR (after all, there is no viewport on the server), and we get the wireframe rendering its content at the top. This is very unfortunate 😅

When I enter production, I don't see the SSR part. There's no `data-calypso-ssr="true"` in `div#wpcom`. That's kind of weird. Maybe it's not using it in production? Or is it so fast I can't even notice it? 🤷🏻 

Anyway... Luckily, the solution is straightforward (if we actually want to change it.) We'll need to delegate the decision of rendering the content at the center of the screen to the CSS realm instead of JS:

```diff
diff --git a/packages/onboarding/src/step-container-v2/components/ContentWrapper/style.scss b/packages/onboarding/src/step-container-v2/components/ContentWrapper/style.scss
index 5165e62bddf..52bfa8cf3f7 100644
--- a/packages/onboarding/src/step-container-v2/components/ContentWrapper/style.scss
+++ b/packages/onboarding/src/step-container-v2/components/ContentWrapper/style.scss
@@ -32,9 +32,11 @@
 	);
 
 	&.center-aligned {
-		justify-content: center;
-		padding-bottom: calc(
-			var( --step-container-v2-content-block-padding ) + var( --step-container-v2-top-bar-height )
-		);
+		@include break-small {
+			justify-content: center;
+			padding-bottom: calc(
+				var( --step-container-v2-content-block-padding ) + var( --step-container-v2-top-bar-height )
+			);
+		}
 	}
 }
diff --git a/packages/onboarding/src/step-container-v2/wireframes/CenteredColumnLayout/CenteredColumnLayout.tsx b/packages/onboarding/src/step-container-v2/wireframes/CenteredColumnLayout/CenteredColumnLayout.tsx
index d4d29a6167a..e676dd7e1ab 100644
--- a/packages/onboarding/src/step-container-v2/wireframes/CenteredColumnLayout/CenteredColumnLayout.tsx
+++ b/packages/onboarding/src/step-container-v2/wireframes/CenteredColumnLayout/CenteredColumnLayout.tsx
@@ -33,7 +33,7 @@ export const CenteredColumnLayout = ( {
 				return (
 					<>
 						<TopBarRenderer topBar={ topBar } />
-						<ContentWrapper centerAligned={ context.isSmallViewport && verticalAlign === 'center' }>
+						<ContentWrapper centerAligned={ verticalAlign === 'center' }>
 							{ heading && <ContentRow columns={ 6 }>{ heading }</ContentRow> }
 							<ContentRow columns={ columnWidth } className={ className }>
 								{ content }

```

Even with that, there's a small layout shift (restart the Calypso server for these changes to take effect since it's SSR), but I think it's caused by...

### Page scroll fixer

The page scroll can be addressed by removing the `padding-bottom` and `min-height` from `.layout.is-section-login`.

### Massive blank space in the "continue as user" block

Can be removed by setting `height: auto` and `margin: 0 auto` in `.continue-as-user`.

tellthemachines committed:
Additional merge-related changes

tellthemachines commented:
Thanks for testing @zaguiini!

> * I see the login form blink upon refresh (not happening in prod)

I can repro this on trunk, so looks like it's the SSR issue you're mentioning above.

> * It's not centered aligned on load. There's a layout shift

Hmm yeah this potentially affects all screens using `CenteredColumnLayout` with vertical centering, right? If that's the case I think it's better to fix it in a separate PR.
> * I don't see the subheader text anymore. Is that expected? We have it in prod.

The ToS moved to the bottom of the screen, as per the updated design. The copy also changed a bit so it's shorter now.

> * Then I see a massive blank space between my user action and the change account escape hatch
> * Finally, the page scrolls when it shouldn't

These two seem to be related, and I can repro them in production. Looks like they're due to a change introduced in #102923 - cc @adamwoodnz 

I can push an override for `isWhiteLogin` screens to remove the height altogether, so "log in with another account" will be right under "continue", but might be best to check with design if that's suitable:

<img width="1032" alt="Screenshot 2025-05-05 at 12 01 48 pm" src="https://github.com/user-attachments/assets/9eb29b5f-953b-40dd-9e2f-a7d917bb52af" />

There's also a bit of layout shift in production when the language notice pops in:

https://github.com/user-attachments/assets/3af1739d-94c4-43cb-b0a2-273d47b81ea5

tellthemachines committed:
Remove continue as user height

adamwoodnz commented:
> These two seem to be related, and I can repro them in production. Looks like they're due to a change introduced in #102923 - cc @adamwoodnz

The massive blank space has always been there from what I can tell. My change actually pulled it up a little (see screenshots on that PR). I think the intention is for it to be positioned near the bottom of a mobile screen. Both reviewers suggested revisiting that layout so I think it's worth discussing with design 👍

tellthemachines committed:
Remove convoluted button widths

tellthemachines commented:
> I don't want us to hack the max-width best practice, but I think that from `599px` to the smallest, the sizes should be the same.

I tried removing the responsive logic altogether so that button max-width is always `327px`. Let me know if that doesn't work!

tellthemachines committed:
Fix scrollbar

tellthemachines commented:
Ok turns out the scrollbar was also being caused by some styles that we don't need on `isWhiteLogin` screens, so I adjusted them to not apply here. Everything should look good now!

@fcoveram would love your opinion on whether the continue as user screen looks good with the min-height removed:

<img width="428" alt="Screenshot 2025-05-06 at 10 19 24 am" src="https://github.com/user-attachments/assets/aeaa35ec-fbd8-4416-ab50-f7e5be0e8d3e" />

<img width="1156" alt="Screenshot 2025-05-06 at 10 53 48 am" src="https://github.com/user-attachments/assets/00868781-1a73-4864-9df9-aa2626092f83" />

And also regarding the adjustments to login button width on mobile!

talldan commented on client/blocks/login/login-header.tsx:
```diff
@@ -0,0 +1,597 @@
+	showContinueAsUser: () => boolean;
+}
+
+export function GetHeaderText( {
```
Nit: The `Get` naming feels unusual for a react component.

Looking at the implementation, maybe it should be a hook (`useHeaderText`) if it needs to use `useTranslate`, or it could be a regular function that has `translate` as an argument.

tellthemachines replied:
yeah I went back and forth on this. It can't be a hook because I need to call it inside the Login class component, and if I made it a regular function typescript got really sulky about the `translate` calls that use args because it thinks they're not outputting a string. In the end this was the simplest solution. I could rename it to `HeaderText` which might be slightly better?

talldan commented on client/blocks/login/login-header.tsx:
```diff
@@ -0,0 +1,597 @@
+	twoFactorEnabled,
+	currentQuery,
+	showContinueAsUser,
+}: {
+	isSocialFirst: boolean;
+	twoFactorAuthType: string | null;
+	isManualRenewalImmediateLoginAttempt: boolean;
+	twoStepNonce: string | null;
+	socialConnect: boolean;
+	linkingSocialService: string;
+	action: string;
+	privateSite: boolean;
+	oauth2Client: { title: string; icon: string } | null;
+	isWooJPC: boolean;
+	isFromMigrationPlugin: boolean;
+	isJetpack: boolean;
+	isWCCOM: boolean;
+	isFromAkismet: boolean;
+	isFromAutomatticForAgenciesPlugin: boolean;
+	isGravPoweredClient: boolean;
+	wccomFrom: string;
+	twoFactorEnabled: boolean;
+	currentQuery: Record< string, string >;
+	showContinueAsUser: () => boolean;
+} ): JSX.Element {
```
From what I can tell, it looks like the properties here are a subset of the properties in `LoginHeaderProps`.

Caveat that I'm not TS expert, so there might be another option, but I think some of the boilerplate could be reduced by using intersection types - https://www.typescriptlang.org/docs/handbook/2/objects.html#intersection-types.

So the highlighted code were defined as `HeaderTextProps`, `LoginHeaderProps` could be defined as something like: 
```ts
type LoginHeaderProps = HeaderTextProps & { 
    isGravPoweredLoginPage: boolean,
   // ... and any remaining types not already in `HeaderTextProps`
};
```

tellthemachines replied:
Oooh yes good idea, I'll try that!

tellthemachines replied:
Ok so I changed this to a regular function and inlined the types, so shouldn't need to do this now.

talldan replied:
Nice, looks much tidier now without the fragments 👍

talldan commented on client/blocks/login/login-header.tsx:
```diff
@@ -0,0 +1,597 @@
+} ): JSX.Element {
+	const translate = useTranslate();
+
+	let headerText = <></>;
```
Is the use of fragment throughout to satisfy the type checker?

It looks like you can import `TranslateResult` from 'i18n-calypso' and use that as the return type. `useTooltip` is the example that I found that does this.

tellthemachines replied:
Ah I see, maybe that could solve the need for this function to be a component 🤔 thanks for the tip!

talldan commented on client/blocks/login/login-header.tsx:
```diff
@@ -0,0 +1,597 @@
+	let headerText = <></>;
+
+	if ( isSocialFirst ) {
+		headerText = <>{ translate( 'Log in to WordPress.com' ) }</>;
```
Could it be possible to early return the translation result instead of assigning to `headerText`. It'd be a minor optimization, but it'd mean not having to check every case at run time.

tellthemachines replied:
I chose to leave the conditional logic as it was because some cases override others, and I didn't want to accidentally break a login screen that I don't even know about (I think I've got most of them by now, but keep finding new instances so can't be 100% sure 😅)

talldan commented on client/blocks/login/login-header.tsx:
```diff
@@ -0,0 +1,597 @@
+			}
+		}
+
+		if ( isJetpackCloudOAuth2Client( oauth2Client ) || isA4AOAuth2Client( oauth2Client ) ) {
```
I guess it was like this before and we don't want to refactor too much, but the number of permutations/conditions here is pretty mind boggling!

How is it possible to test all these 😱 

I guess the answer is that it isn't. 😄

tellthemachines replied:
😂 😭

talldan reviewed (commented):
Left a few comments, but I ran out of time to test or fully review.

fcoveram reviewed (approved):
> would love your opinion on whether the continue as user screen looks good with the min-height removed:

> And also regarding the adjustments to login button width on mobile!

The page alignment and button width look good on both viewports.

Regarding the language banner, I noticed that ticket #100426 was about moving it to the login page. The screenshots attached in the comments don’t include this screen, but if it’s part of the flow, then I would keep the latest implementation untouched.

tellthemachines committed:
Improve syntax

jsnajdr commented:
When I'm already logged in, I see the `<ContinueAsUser>` screen first and its layout is broken. The "Continue" button and the "Not you?" text under it overlap, no matter how I resize my screen. The sentence after, "Log in with another account" also has an unnecessary line break.

<img width="611" alt="Screenshot 2025-05-07 at 13 59 29" src="https://github.com/user-attachments/assets/caf77d8c-6ed4-415d-a357-82b3a38e14b4" />

This PR also could benefit from being merged as a series of smaller PRs. For example, what if the `blocks/login/index.js` component was first refactored from a class component from functional? That could make further updates more pleasant.

tellthemachines commented:
> The "Continue" button and the "Not you?" text under it overlap, no matter how I resize my screen. The sentence after, "Log in with another account" also has an unnecessary line break.

Thanks for testing @jsnajdr! Weirdly, I can't reproduce that but I see how it _might_ happen because the "Not you?" is absolutely positioned in its container that now has auto height. I'll remove the positioning because it's no longer necessary without the fixed height.

> For example, what if the blocks/login/index.js component was first refactored from a class component from functional? That could make further updates more pleasant.

I'm not convinced that would help much here 😅 because what makes the code so unpleasant is the amount of slightly different use cases this component is catering to. What would really improve maintainability would be to unify the design of all these login screens so we could remove all the convoluted logic for each specific instance. Based on [this comment](https://github.com/Automattic/wp-calypso/pull/102353#issuecomment-2808995385) I'm hopeful we'll get there eventually!

tellthemachines committed:
Remove positioning from "Not you"

adamwoodnz commented:
I'm getting a crash on the [Woo login screen](http://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%2F%3Fresponse_type%3Dcode%26client_id%3D50916%26state%3D69c7f83c44d9438e013f60308e0d8369725e57fa68d18b558d8b41743976985b%26redirect_uri%3Dhttps%253A%252F%252Fwoocommerce.com%252Fwc-api%252Fwpcom-signin%253Fnext%253D%25252F%26blog_id%3D0%26wpcom_connect%3D1%26wccom-from%26calypso_env%3Dproduction%26from-calypso%3D1&client_id=50916&signup_url=%2Fstart%2Fwpcc%2Foauth2-user%3Foauth2_client_id%3D50916%26oauth2_redirect%3Dhttps%253A%252F%252Fpublic-api.wordpress.com%252Foauth2%252Fauthorize%252F%253Fresponse_type%253Dcode%2526client_id%253D50916%2526state%253D69c7f83c44d9438e013f60308e0d8369725e57fa68d18b558d8b41743976985b%2526redirect_uri%253Dhttps%25253A%25252F%25252Fwoocommerce.com%25252Fwc-api%25252Fwpcom-signin%25253Fnext%25253D%2525252F%2526blog_id%253D0%2526wpcom_connect%253D1%2526wccom-from%2526calypso_env%253Dproduction%2526from-calypso%253D1)

![Screenshot 2025-05-09 at 11 40 03 AM](https://github.com/user-attachments/assets/271c9e8b-39ee-43fd-9106-096c427a0a22)

Just me?

tellthemachines commented:
> Just me?

I can repro. Looking into it!

tellthemachines committed:
Update `showContinueAsUser` to optional boolean in getHeaderText

tellthemachines commented:
Fixed, thanks for catching that @adamwoodnz!

adamwoodnz commented:
> Fixed, thanks for catching that @adamwoodnz!

You're welcome! Works for me. Just found something else though sorry. If you click the 'Email me a login link' button the content on the magic login screen isn't centered:

![calypso localhost_3000_log-in_redirect_to=https%3A%2F%2Fpublic-api wordpress com%2Foauth2%2Fauthorize%2F%3Fresponse_type%3Dcode%26client_id%3D50916%26state%3D69c7f83c44d9438e013f60308e0d8369725e57fa68d18b558d8b41743976985b%26redirect_uri%3D](https://github.com/user-attachments/assets/a23efd4c-43a8-41bd-876f-0742fa5fc36b)

adamwoodnz commented:
...and the Wordpress.com version has a truncated layout

![calypso localhost_3000_log-in_link_signup_url=%2Fstart%2Faccount%2Fuser-social(5 1440x900)](https://github.com/user-attachments/assets/c7c4a2ef-2531-450f-86a3-07e24d0f4480)

tellthemachines commented:
Oh, the Woo magic login screen _also_ uses `isWhiteLogin` styles 🤦 

Ok I'll have to check all the magic logins and work out how to isolate those styles.

tellthemachines commented:
@adamwoodnz do you know how to test [the `isWooJPC`](https://github.com/Automattic/wp-calypso/blob/update/login-screens-content/client/login/magic-login/index.jsx#L1295) and [the `isStudioAppOAuth2Client`](https://github.com/Automattic/wp-calypso/blob/update/login-screens-content/client/login/magic-login/index.jsx#L1335) variants of the login screen? (is the studio app one even active anymore?) 

I think the problem with the Woo screen is that it shouldn't be getting the `is-white-login` class applied, because it isn't a `isWhiteLogin` screen. Ideally we'd have that class apply conditionally. I can do it for the [general case](https://github.com/Automattic/wp-calypso/blob/update/login-screens-content/client/login/magic-login/index.jsx#L1367) anyway, but would be nice if we could also clean up the others.

tellthemachines committed:
Fix use of `is-white-login` in magic login component

tellthemachines committed:
Fix height in magic login

tellthemachines commented:
Ok all the styling issues should be fixed now. I checked all the screens _that I know of_ and everything seems correct.

adamwoodnz commented:
> @adamwoodnz do you know how to test [the `isWooJPC`](https://github.com/Automattic/wp-calypso/blob/update/login-screens-content/client/login/magic-login/index.jsx#L1295) and [the `isStudioAppOAuth2Client`](https://github.com/Automattic/wp-calypso/blob/update/login-screens-content/client/login/magic-login/index.jsx#L1335) variants of the login screen? (is the studio app one even active anymore?)

I don't sorry. https://developer.wordpress.com/studio/ -> Get Started takes me through the typical WordPress.com login flow.

Tested, and all the magic logins LGTM. Great work on this, there are so many variations, I know how tricky it is 😅

andrewserong commented:
I've only tested a few variations of the login screen, but all the Dotcom-ish ones are all looking good to me so far! The one issue I ran into was with the A4A login link, using the preview link in [the comment above](https://github.com/Automattic/wp-calypso/pull/102811#issuecomment-2826262282). I.e. [this A4A link](https://calypso.live/?image=registry.a8c.com/calypso/app:build-143510&env=a8c-for-agencies) in an incognito browser.

| Other PRs | This PR |
| --- | --- |
| <img width="1177" alt="image" src="https://github.com/user-attachments/assets/0a6ce229-6ece-481b-9ce2-3c70fdef107f" /> | <img width="1269" alt="image" src="https://github.com/user-attachments/assets/c9eff1e2-b924-4394-9acd-1387780e6317" /> |

In practice, though, I wasn't sure if anyone would ever usually see that screen? In the "real" A4A site in production I see this login screen instead: 

<img width="1402" alt="image" src="https://github.com/user-attachments/assets/1b085c29-57f6-4605-9dc5-b5c288246c11" />

tellthemachines commented:
> In practice, though, I wasn't sure if anyone would ever usually see that screen? In the "real" A4A site in production I see this login screen instead:

Yeah, I can't work out how to reach this screen in prod. What's happening style-wise is a weird intersection of conditions, where this screen is within the scope of `isWhiteLogin` but its component parts aren't really adequate for that design. Also, the fact that "Login via the mobile app" link is displaying but errors when clicked makes me suspect this screen is just some weird artifact born of over-complex code 😅

tellthemachines commented:
Ok so that login can be accessed in prod through this URL: https://agencies.automattic.com/log-in?redirect_to=wat

However, trying to actually log in from there isn't working for me 😕 

Also, that URL doesn't work without the query string: https://agencies.automattic.com/log-in 

which sort of indicates that it probably shouldn't exist. Given the length of this PR, it might be better to remove that login screen altogether as a separate PR.

tellthemachines committed:
Update header condition to isWhiteLogin

tellthemachines commented:
As I was looking at this I realised the condition for rendering the updated header was slightly inconsistent with the rest of the PR so I updated it to use `isWhiteLogin` instead of `isSocialFirst`. (Also because `isSocialFirst` should be removed altogether in a follow-up)

andrewserong commented on client/blocks/login/login-form.scss:
```diff
@@ -15,19 +15,25 @@ form {
+
 	form {
 		max-width: 400px;
+		margin: 0;
```
The default state (i.e. testing with the A4A link) is looking better after the latest commit. There still appears to be an alignment issue that I think might be caused by this rule? Apologies, I haven't had a deep look through the code change here so the answer might be obvious, but does this rule to clear out the margins on the form always need to apply? From the look of the other variations of the login form, I assume the clearing out is meant to happen when the social buttons are available, so I wondered if the `margin: 0` could be moved down to the `&.is-social-first` block below? Or do other variations outside of `is-social-first` also need that 0 margin? 🤔

Here's what I'm seeing:

| With this rule, form is left aligned | Without this rule form is centered as before |
| --- | --- |
| <img width="1215" alt="image" src="https://github.com/user-attachments/assets/e930952f-4073-442c-9436-d634c3ef174f" /> | <img width="1214" alt="image" src="https://github.com/user-attachments/assets/c561cba7-7714-47e8-a2a4-bee3f4f7aa9e" /> |

tellthemachines replied:
This is complicated 😅 

I could push a CSS fix for that alignment issue. If so, it would be better to target that specific screen (say by using the `theme-a8c-for-agencies` body class) than change the scope of the existing styles. This is for two reasons:

* `is-social-first` is attached to an obsolete feature flag that I plan to remove in a future PR;
* I believe this login screen shouldn't exist, so I also plan to remove it in a future PR. I described the reasoning for that [here](https://github.com/Automattic/wp-calypso/pull/102811#issuecomment-2870384960).

Because the screen will be removed, I thought this wasn't worth fixing, but I'm not too fussed either way. What do you think?

andrewserong replied:
Thanks for laying all that out!

> I believe this login screen shouldn't exist, so I also plan to remove it in a future PR.

Removing unused screens sounds great to me 👍

> Because the screen will be removed, I thought this wasn't worth fixing, but I'm not too fussed either way. What do you think?

If we're sure that there's no way that folks will land on this screen in a production environment (and it seems we haven't found anywhere) then I agree, it doesn't sound worth fixing, especially since it'll be removed shortly, anyway. My only hesitation was because I don't have a solid grasp of all the ways in which we're linking to or using the login screen.

But all that said, my thinking is:

* If all the prod use cases we're aware of are looking good, and:
* The base `/log-in` with no params looks good (as in the case of wordpress.com / running Calypso locally)

Then we should be fine. And since we're aware of this alignment issue, if we encounter it anywhere out in the wild, then it'll likely be a) an incredibly low priority login flow, and b) we've finally learned we're it's still used 😅

tellthemachines replied:
I just pushed a fix in any case. I'll remove it once the actual screen is removed.

tellthemachines replied:
> * I believe this login screen shouldn't exist, so I also plan to remove it in a future PR. I described the reasoning for that [here](https://github.com/Automattic/wp-calypso/pull/102811#issuecomment-2870384960).

I brought this up with the A4a dev team and they [created an issue](https://linear.app/a8c/issue/A4A-873/a4a-login-screen-shows-empty-screen-when-visited-manually) to fix it. I'll keep an eye on it as I imagine there'll be some design work involved if they want to preserve the screen, so it would be good to make sure that work doesn't add further complexity to our login logic.

tellthemachines committed:
Fix weird a4a login screen

andrewserong commented on client/blocks/login/index.jsx:
```diff
@@ -340,415 +331,20 @@ class Login extends Component {
-			if ( isA4AOAuth2Client( oauth2Client ) ) {
-				headerText = translate(
-					'Howdy! Log in to Automattic for Agencies with your WordPress.com{{nbsp/}}account.',
+	renderLoginFromSignupNotice() {
```
Typo: should this function be `renderLoginFormSignupNotice`? 😄

tellthemachines replied:
Probably? I just copied it as it was from `client/blocks/login/login-form.jsx`.

andrewserong reviewed (commented):
I've been testing through all the various login forms that have been linked to so far, and looking good as far as I can tell!

✅ Looking good for WordPress.com
✅ Looking good for Woo.com
✅ Looking good for Automattic partner portal at hosts.automattic.com
✅ Looking good for WP Cloud
✅ Looking good for Akismet
✅ Looking good for Gravatar
✅ Looking good for WP Job Manager
✅ Looking good for Blaze Pro
✅ Looking good for Jetpack

I've only made a start in reading the actual code change, but I'll continue on tomorrow if no-one beats me to it 🙂

Edit: I've added the String Freeze label since it looks like this PR adds one new untranslated string.

jsnajdr commented:
> what makes the code so unpleasant is the amount of slightly different use cases this component is catering to. What would really improve maintainability would be to unify the design of all these login screens so we could remove all the convoluted logic for each specific instance.

Yes, the current situation where every product (Blaze Pro, CrowdSignal, flavors of Woo, Gravatar) can do almost unlimited customizations is clearly untenable. And given the login flow complexity (many social flows, many variants of 2FA), I doubt they all work 100% correctly.

What we should do is to offer just a handful of focused customization points: color, logo, header text and font. The rest of the UI would be unified. That's something that's possible to maintain and test.

Inspired by reviewing this PR, I already managed to do three login cleanups: #103251, #103230 and #103323.

tellthemachines commented:
> Inspired by reviewing this PR, I already managed to do three login cleanups: #103251, #103230 and #103323.

Could you please save any further cleanup until after this PR is merged? Merge conflicts aren't fun to deal with.

tellthemachines committed:
Remove privateSite mentions

tellthemachines committed:
Reapply changes from #103213

tellthemachines committed:
Fix typo

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/17464924).

andrewserong reviewed (commented):
This is still testing well for me, and I like that the refactor splits out `getHeaderText` from the LoginHeader, and that the LoginHeader is extracted from `Login`. These feel like good steps forward to me 👍. `getHeaderText`'s signature is pretty long and I could imagine it being easy to accidentally make a mistake on ordering of the params. Not a blocking comment, but if we were going to use this function in additional places, I'd suggest using an object as a param so that we don't need to worry about the positioning of the params.

I haven't been able to find any issues with each of the different logins I've attempted with this PR checked out, though it's very hard to be sure that I've covered everything, but from my perspective I think this change is looking pretty good to land. And we've certainly been covering the main known use cases a lot in our testing (WP.com, Woo, Jetpack, WPCloud, A4A, etc).

@jsnajdr I see you've been reviewing this PR — are you still looking through it, or do you think it's okay to ship? I'm aware it looks like there've been other discussions about ideas for how to improve the login code, so wasn't sure where this sits in priorities. IMO it'd be nice to get the design improvements from this PR in.

> What we should do is to offer just a handful of focused customization points: color, logo, header text and font. The rest of the UI would be unified. That's something that's possible to maintain and test.

That does like a good idea for the long-term to me!

tellthemachines commented:
> `getHeaderText`'s signature is pretty long and I could imagine it being easy to accidentally make a mistake on ordering of the params. Not a blocking comment, but if we were going to use this function in additional places, I'd suggest using an object as a param

The intention is not to make it reusable 😅 as this header text is very specific to its context. I only extracted the function so I could keep that horrible pile of conditions unchanged while applying it both to the redesigned (`isWhiteLogin`) login screens and all the other ones (which require extra markup that `LoginHeader` provides). Ideally we'll work to simplify the conditions in follow ups so that function becomes less gnarly.

This PR has been going on for a while and I think it's in a pretty stable state now, so if no one is opposed I'm going to go ahead and merge it within the next few hours. Otherwise the chance of further merge conflicts is high and it's becoming a bit of a time sink.

andrewserong reviewed (commented):
Just re-tested, and I think there's one more area of styling that needs resolving. When I log in with an account that uses 2FA, the screen after submitting my username and passwords has alignment that's a bit off:

<img width="1447" alt="image" src="https://github.com/user-attachments/assets/44af31d6-b9d0-4aa4-b52a-5b28a0e07953" />

It looks similar to the `margin: 0` issue we were looking at earlier?

<img width="1620" alt="image" src="https://github.com/user-attachments/assets/09b2605a-a65d-4913-b20b-fcbced21bc9c" />

tellthemachines commented:
> When I log in with an account that uses 2FA, the screen after submitting my username and passwords has alignment that's a bit off:

Oh weird, I was sure I'd tested that. Thanks for spotting it!

tellthemachines committed:
Fix margin for 2fa screen

tellthemachines commented:
Ok I've fixed the margins on the 2fa screen, but I'm not really happy with the fix as I had to leverage the `is-social-first` classname attached to the form element. The problem with this is that the 2fa screen also technically falls under the `isSocialFirst` umbrella (it has the classname on its `main` element), though the fix works because it renders a different form element from the other screens, which doesn't have that class.

I opted for this because I don't want to further complicate and increase testing scope for this PR. I hope to follow up with another PR to untangle `isSocialFirst`.

jsnajdr commented:
> Could you please save any further cleanup until after this PR is merged? Merge conflicts aren't fun to deal with.

@tellthemachines Oh, I apologize for being so unhelpful while attempting to be helpful 🙁 

> I see you've been reviewing this PR — are you still looking through it, or do you think it's okay to ship?

@andrewserong I got to review it again only now, after merge. Given that the login page is such a monster, it's really hard to determine if such a big change is OK to ship. I'm afraid we'll need to do a few post-merge fixes as problems are discovered.

Here's the first one: the Woo Core Profiler Lost Password Flow (whatever that is 🙂) is broken. Go to:
```
https://wordpress.com/log-in?from=woocommerce-onboarding&lostpassword_flow=true
```
both logged in and logged out. The logged out page is broken in this way:
<img width="973" alt="Screenshot 2025-05-14 at 8 59 17" src="https://github.com/user-attachments/assets/62ea3394-dc05-4855-83ef-2860188663a2" />

There's a `WooCoreProfilerMasterbar` component conflicting with the WordPress.com logo header. Also there should be a different heading, with an explanatory text below. This is the version before this PR:

<img width="960" alt="Screenshot 2025-05-14 at 9 16 15" src="https://github.com/user-attachments/assets/7eee446e-8530-430d-8dd8-c4136b1ff604" />

The logged-in version, with `ContinueAsUser`, is broken in a similar way: double header, wrong heading.

<img width="975" alt="Screenshot 2025-05-14 at 9 00 13" src="https://github.com/user-attachments/assets/c7567ec6-72d2-44ae-9de6-7c0e5f7d4d88" />

Additionally, the `ContinueAsUser` UI is broken, with cut "Continue" button and displaced "Not you" text. But interestingly this i a pre-existing bug, this part was already broken.

jsnajdr commented on client/blocks/login/index.jsx:
```diff
@@ -337,413 +328,20 @@ class Login extends Component {
-					);
-				}
-			}
-		} else if ( isWooJPC ) {
```
When the flow is `isWooJPC` then `isWhiteLogin` is false and this `LoginHeader` component is not even rendered. That's one reason why this flow (including "lost password") has incorrect header and sub-header.

jsnajdr commented on client/blocks/login/login-header.tsx:
```diff
@@ -0,0 +1,532 @@
+		const isLostPasswordFlow = currentQuery.lostpassword_flow === 'true';
+		switch ( true ) {
+			case isLostPasswordFlow:
+				headerText = translate( "You've got mail" );
```
There are missing `break` statements in this `switch`. The `isLostPasswordFlow` condition falls through to the following condtions and the `headerText` is overwritten.

Personally I would rewrite these `switch ( true )` statements to a regular `if ... else if ... else`.

jsnajdr reviewed (commented):
We have one more trouble with the `isWhiteLogin` value: it's assigned at several places: login controller, logged-out layout (masterbar display is determined there), and each time the condition is a bit different 🤦

jsnajdr commented:
One more broken page, again the WooJPC flow. Double masterbar, and misaligned icons on the social login buttons:

<img width="944" alt="Screenshot 2025-05-14 at 13 53 31" src="https://github.com/user-attachments/assets/397c92ee-0d90-4f88-8978-09faebd3c288" />

tellthemachines commented:
Thanks for reviewing and testing @jsnajdr! I suspect those Woo screens have been broken since #102353, when I updated `isWhiteLogin` screens to use `Step.TopBar` 😬  The only real changes I made to `isWhiteLogin` in this PR were to remove Gravatar from the condition, so I'm guessing the Woo screens were already included in the condition back then.

Question: how did you get to those screens? If I go to `woo.com` and try to login from there, I get different layouts from those (which are displaying correctly, because I knew to test them 😅 ) but it looks like those screens are reached by a different path? 

> Given that the login page is such a monster, it's really hard to determine if such a big change is OK to ship. I'm afraid we'll need to do a few post-merge fixes as problems are discovered.

This was absolutely what I was expecting when I merged the PR! Given it had been open for 3 weeks and had a substantial amount of testing I was fairly sure nothing major was broken, but any large change to a piece of spaghetti like this is bound to break _something_. I'll follow up with some fixes 😄

jsnajdr commented:
> Question: how did you get to those screens?

I was reading the code, inspected the various conditions and then tried to reconstruct the actual URL to match that condition. I didn't test the full flows that start somewhere on woocommerce.com or inside a self-hosted site's WP Admin.

With that approach, it's possible that some URLs don't really exist in practice. For example, it's possible that the real flow sends you to `/log-in/jetpack?from=woocommerce-core-profiler` and never to `/log-in?from=woocommerce-core-profiler`. The combination "no Jetpack + Woo Core" doesn't occur in real use and yet we might be testing and fixing it.

adamwoodnz commented:
I'm not 100% sure it's related to this PR but I discovered an issue on the Blaze Pro 2FA screen where the logos are clashing in the header:

![wordpress com_log-in_lostpassword_redirect_to=https%3A%2F%2Fpublic-api wordpress com%2Foauth2%2Fauthorize%3Fresponse_type%3Dcode%26client_id%3D99370%26state%3D6e545c7a9932a0a612817faa07992930afae8229e4b11a219e9db95944038cb9%26redirect_u (2)](https://github.com/user-attachments/assets/7de89510-1b41-4650-8618-68460de3d1ed)
